### PR TITLE
Mongodb for mpapi

### PR DESCRIPTION
**Pull request by ChrisFajardo-TRI**
2 main changes

- environment variables for the piro module moved to piro.settings
  - used pydantic because that's the new hotness on this type of stuff (loading/validation handling) and integrates well with the web app, but can reimplement without it
  - easier to mock/manage as an import
  - I plan to move the "use_cache_" type of arguments in SynthesisRoutes.__init__() here as well because they're more general global kind of settings

- new piro.mprester file that holds the MongoMPRester and a get_mprester() factory function to choose which one to use
  - overrode the MPRester.query() to use the mongo db
  - added tests which ensure the results are the same 

While I'm working with mongo, I plan to add another commit that simplifies the epitaxial and similarities cache db code as well

### CODE REVIEW

**Comment by ChrisFajardo-TRI** [comment]
latest commit is that, I moved epitaxy and similarity cache code to utils and tried to make smaller functions that are more reusable

also moved the cache config bools to environment variables

CACHE_TYPE=no_cache|file_cache|mongo_cache
USE_MAPI_DB=0|1
MONGODB_URI

I wonder if we can simplify this or do we need the flexibility. basically, if MONGO_URI is there, try to use it for epitaxy/similarity and MP API?

File: piro/mprester.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        criteria,
+        properties,
+        chunk_size=500,
+        max_tries_per_chunk=5,
+        mp_decode=True,
+    ):
+        """
+        Arguments are the same as superclass. See the superclass's docs for info.
+
+        TODO is this a correct assumption?
```
Comment (JosephMontoya-TRI) [inline piro/mprester.py:48]: Yep, task_id is material_id.

**Review by JosephMontoya-TRI — APPROVED** [review]
This all looks solid to me, addressed one TODO.

**Comment by JosephMontoya-TRI** [comment]
> I wonder if I should simplify this or do we need the flexibility. basically, if MONGO_URI is there, try to use it for epitaxy/similarity and MP API?

I don't think we need the flexibility if you want to simplify, honestly I think our deployed version is the only thing that will have access to these anyway.

File: piro/mprester.py (hunk 1 above)
Comment (ChrisFajardo-TRI) [reply piro/mprester.py]: thanks, removed the comment

**Comment by ChrisFajardo-TRI** [comment]
> > I wonder if I should simplify this or do we need the flexibility. basically, if MONGO_URI is there, try to use it for epitaxy/similarity and MP API?
> 
> I don't think we need the flexibility if you want to simplify, honestly I think our deployed version is the only thing that will have access to these anyway.

I'll keep the current environment variables for now. reason being it looks like the results differ between the mongo epitaxy|similarity cache and the locally calculated ones.  This appears to have existed before this PR, so I'll merge this.
I may try to pinpoint the cause of the difference, and if it's in the piro code, create a new PR to fix

**Comment by JosephMontoya-TRI** [comment]
The difference between the cache may come from the use of the conventional cell in the cache.  It's slightly different than what was presented in the paper, but I think we may want to go ahead and bite the bullet and do the same thing in piro.  What do you think @MuratAykol-TRI?

**Comment by MuratAykol-TRI** [comment]
I think it's fine and differences should be minute so as long as we keep the option to switch back to primitive, in case anyone wants to get numbers consistent with the paper.

**Pull request merged**